alg/teca_2d_component_area.cxx

```diff
@@ -0,0 +1,352 @@
+#include "teca_2d_component_area.h"
+
+#include "teca_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_metadata.h"
+#include "teca_cartesian_mesh.h"
+
+#include <algorithm>
+#include <iostream>
+#include <deque>
+#include <set>
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+namespace {
+
+template <typename label_t>
+label_t get_max_label_id(unsigned long n, const label_t *labels)
+{
+    label_t max_label_id = std::numeric_limits<label_t>::lowest();
+    for (unsigned long i = 0; i < n; ++i)
+    {
+        label_t label = labels[i];
+        max_label_id = label > max_label_id ? label : max_label_id;
+    }
+    return max_label_id;
+}
+
+// visit each node in the mesh, the node is treated as a cell in
+// the dual mesh defined by mid points between nodes. the area of
+// the cell is added to the corresponding label. this formulation
+// requires a layer of ghost nodes.
+//
+// A = rho^2(cos(phi_0) - cos(phi_1))(theta_1 - theta_0)
```

Comment:
In terms of my comment below, you might consider formulating this instead as

```suggestion
// A = rho^2(phi_1 - phi_0)(theta_1 - theta_0)cos(0.5*(phi1 + phi0))
```

where here `phi = y*rad_per_deg` instead of `phi = (90-y)*rad_per_deg`